cpp/ql/lib/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasAnalysis.qll

```diff
@@ -266,6 +266,20 @@ private predicate operandReturned(Operand operand, IntValue bitOffset) {
   bitOffset = Ints::unknown()
 }
 
+pragma[nomagic]
```

Comment:
can you explain to a newbie what does this incantation do? :slightly_smiling_face:

Reply:
I can! "Magic" is a datalog-specific optimization that is implemented by the QL compiler. This optimization gathers "context" from places where the predicate is called, and pushes that context into the predicate body, in order to make the number of tuples in the predicate smaller (and thus the evaluation faster). `pragma[nomagic]` disables that optimization for this predicate.

Normally, when we factor out some joins into their own predicate, we do this because we want those joins to happen in isolation. If we allow magic in that predicate, chances are that the compiler will just insert those additional joins (that we wanted to prevent) back into this new predicate.

So by marking the predicate as `nomagic` we're saying "I pulled out these joins, and I don't want the compiler to insert additional joins". Does that make sense?